### DISCUSSION TITLE
Being able to target elements with data attributes

### DISCUSSION THREAD

**Discussion by Sparkmasterflex** [facebook/stylex#380] [Q&A] [answered]
### Describe the feature request

I'm working on a project where we're combining StyleX and Radix-UI and their [Select](https://www.radix-ui.com/primitives/docs/components/select#installation) component has a `disabled` [state](https://github.com/radix-ui/primitives/blob/c31c97274ff357aea99afe6c01c1c8c58b6356e0/packages/react/select/src/Select.tsx#L1235-L1236) for options. It puts this value as a `aria-disabled` and a `data-disabled` attributes and then style with css off of this:

<img width="517" alt="Screenshot 2024-01-25 at 3 14 14 PM" src="https://github.com/facebook/stylex/assets/726129/0c3b8c37-9815-4759-9e95-e547c8f28176">

```css
.SelectItem[data-disabled] {
  color: var(--mauve-8);
  pointer-events: none;
}
```

I think because it's not an actual `<option>` that they're rendering, it wouldn't be valid for them to use the actual `disabled` attribute that is put on form fields and such.

I've been looking through docs and code for StyleX and I can't find any way to do this

**Comment by nmn** [comment, marked as answer]
This pattern isn't officially supported, but you can use `:is([data-disabled])` where you would normally use `:hover` to get around this problem for now.

We don't recommend this pattern in general and prefer applying styles conditionally instead. Using attribute selectors leads to a duplicate CSS rule where an existing className would've sufficed.

**Reply by Sparkmasterflex** [reply]
Bam! that worked thank you

**Comment by hexcowboy** [comment]
@nmn is this going to be supported at some point? it feels strange to use a pattern that is not recommended. it also seems standard for headless ui libraries to use data attributes for state.

for others stumbling upon this - discussion #509 was helpful specifically for react-aria

**Reply by nmn** [reply]
> it feels strange to use a pattern that is not recommended.

The headless ui libraries are using this pattern so we can't help it. Using `:is()` as a workaround is the current recommendation. It *is* a workaround, but the correct solution would depend on changes to component libraries.

The reason against attribute selectors is that they increase the amount of CSS needed for no reason. These component libraries are using JS to set attributes anyway. They could instead set classNames conditionally instead.

**Reply by tom-sherman** [reply]
Sorry to ressurect this thread, just want to get something clarified.

Both `:is([data-disabled])` and simply `[data-disabled]` seem to work. Aside from the native CSS differences, the only difference is the stylex layer priority I assume? Claude suggests that `is()` is +40, bare attribute selector +3000 because it's classed as an unknown selector.

Would it be recommended to use `is()` still or should the bare attribute selector be used?

To clarify, I understand this isn't idiomatic stylex but this is a pattern that's common in Radix and has continued into the Base UI rewrite so it's going to be a pattern that continues.

Personally I'm working on a design system that uses Base UI and as mentioned above idiomatic Base UI usage is to style based off of attribute hooks. I'm just trying to reconcile these two different idioms as we're about to adopt stylex.

**Reply by chbybnwr** [reply]
what would be the idiomatic stylex anyway?

**Reply by tom-sherman** [reply]
Drive a conditional style from state: `isDisabled && styles.isDisabled`. This is what you'd do naturally when you own the state, but it's not totally natural when the component is uncontrolled.

**Reply by chbybnwr** [reply]
You can pass a function to the render prop of Base UI components to access their states.
For example you can do this with Base UI button instead of using `data-disabled`
```tsx
import { Button as BaseButton } from '@base-ui/react/button'
import * as stylex from '@stylexjs/stylex'

export function Button(props: BaseButton.Props) {
  return (
    <BaseButton
      {...props}
      render={(props, state) => (
        <button
          {...props}
          {...stylex.props(
            state.disabled && styles.disabled,
          )}
        />
      )}
    />
  )
}

const styles = stylex.create({
  disabled: {
    color: 'gray',
  },
})
```

References:
https://base-ui.com/react/handbook/composition#render-function
https://base-ui.com/react/components/button#api-reference

**Reply by tom-sherman** [reply]
It's a good solution (and I may adopt it tbh!) but I'd describe it as a workaround. It's not the way the library wants you to style it's components. Base UI even call this out as an escape hatch of an escape hatch.

If we're talking about idiomatic styling within Base UI (and Radix) then this doesn't fit.

**Reply by nmn** [reply]
We are intentionally not supporting data attributes as a first class feature as it encourages a pattern that will lead to growing CSS. A lot of the inconveniences in StyleX are know but intentional in order to result in the smaller, most reusable CSS possible.

We have escape hatches like using `:is([data-*])` etc, for the rare situations it may be unavoidable, but if you're able to apply a data attribute on an element, you should also be able to apply a style conditionally and that way you don't need any special CSS just to select that data attribute.

I understand that many component libraries expect you to query data attributes so I'll look into making the escape hatch work a bit more reliably, but as of right now I don't think we should make it an official API.
